sled-storage/src/resources.rs

```diff
@@ -0,0 +1,187 @@
+/// of this state can be retrieved by requests to the `StorageManager` task
+/// from the [`crate::manager::StorageHandle`]. This state is not `Sync`, and
+/// as such does not require any mutexes. However, we do expect to share it
+/// relatively frequently, and we want copies of it to be as cheaply made
+/// as possible. So any large state is stored inside `Arc`s. On the other
+/// hand, we expect infrequent updates to this state, and as such, we use
+/// [`std::sync::Arc::make_mut`] to implement clone on write functionality
+/// inside the `StorageManager` task if there are any outstanding copies.
+/// Therefore, we only pay the cost to update infrequently, and no locks are
+/// required by callers when operating on cloned data. The only contention here
+/// is for the refrence counters of the internal Arcs when `StorageResources`
+/// gets cloned or dropped.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct StorageResources {
+    // All disks, real and synthetic, being managed by this sled
+    pub disks: Arc<BTreeMap<DiskIdentity, (Disk, Pool)>>,
+}
+
+impl StorageResources {
+    /// Insert a disk and its zpool
+    ///
+    /// Return true if data was changed, false otherwise
+    ///
+    /// This really should not be used outside this crate, except for testing
+    pub fn insert_disk(&mut self, disk: Disk) -> Result<bool, Error> {
+        let disk_id = disk.identity().clone();
+        let zpool_name = disk.zpool_name().clone();
+        let zpool = Pool::new(zpool_name, disk_id.clone())?;
+        if let Some((stored_disk, stored_pool)) = self.disks.get(&disk_id) {
+            if stored_disk == &disk
+                && stored_pool.info.size() == zpool.info.size()
+                && stored_pool.name == zpool.name
+            {
+                return Ok(false);
+            }
+        }
+        // Either the disk or zpool changed
+        Arc::make_mut(&mut self.disks).insert(disk_id, (disk, zpool));
+        Ok(true)
+    }
+
+    /// Insert a disk while creating a fake pool
+    /// This is a workaround for current mock based testing strategies
+    /// in the sled-agent.
+    ///
+    /// Return true if data was changed, false otherwise
+    #[cfg(feature = "testing")]
+    pub fn insert_fake_disk(&mut self, disk: Disk) -> bool {
+        let disk_id = disk.identity().clone();
+        let zpool_name = disk.zpool_name().clone();
+        let zpool = Pool::new_with_fake_info(zpool_name, disk_id.clone());
+        if self.disks.contains_key(&disk_id) {
+            return false;
+        }
+        // Either the disk or zpool changed
+        Arc::make_mut(&mut self.disks).insert(disk_id, (disk, zpool));
+        true
+    }
+
+    /// Delete a disk and its zpool
+    ///
+    /// Return true, if data was changed, false otherwise
+    ///
+    /// Note: We never allow removal of synthetic disks as they are only added
+    /// once.
+    #[cfg(not(test))]
```

Comment:
> If the issue is that "tests want to remove synthetic disks using a normal pathway", I'd ask: then why have the constraint preventing this from happening in #[cfg(not(test))] code? E.g. Why not enable removal of synthetic disks in both cases?

The problem is that synthetic disks don't get recreated, and what I'm trying to test is the behavior of `ensure_using_exactly_these_disks` on a real system where removes actually remove disks from resources. This test is expected to run on any helios system and therefore has to use synthetic disks, and I didn't want to change this [method call](https://github.com/oxidecomputer/omicron/blob/c8db7c88dec4030f29bea0520dea43e81187c8dd/sled-storage/src/manager.rs#L468-L472).

I also didn't really want to take on the responsibility of changing how synthetic disks get created from virtual hardware in this pr and allow them to be removed, as it is already quite large. I think that the current solution is a reasonable compromise and that the real solution is to do as @smklein suggests and fix how we create and use synthetic disks in sled-agent. That would also remove other conditional code related to queueing synthetic disks and be an improvement.

Reply:
Without making any changes to the general synthetic disk issue, how do you feel about my original suggestion? Make this `remove_disk` always exist, and rename the `cfg(test)` one to make it more explicit that it behaves differently (as opposed to the same function name having different behavior in prod vs test)?

Reply:
> Without making any changes to the general synthetic disk issue, how do you feel about my original suggestion? Make this `remove_disk` always exist, and rename the `cfg(test)` one to make it more explicit that it behaves differently (as opposed to the same function name having different behavior in prod vs test)?

If I make this change it means that I will have to inline calls based on configuration inside `ensure_using_exactly_these_disks`, because that's what I'm trying to test with this change. I want to test the production behavior by changing how the test code for remove operates. I don't know a better way to do this. While I could create a separate function for testing `ensure_using_exactly_these_disks` that could lead to unexpected divergence. I think both these solutions (inlining cfg and having a separate test function) are somewhat worse than how I have things now, but I'm open to the name change and inlining of `cfg` at least.

Reply:
I like this a lot. Done in [cfc3ef7](https://github.com/oxidecomputer/omicron/pull/4332/commits/cfc3ef7c732afdfbf476ac4a62b64471c697bfbb)